### PR TITLE
Major update for the Reloop Beatpad : Harder, better, faster !

### PR DESCRIPTION
**Pull request by DJChloe**
Happy New Year 2016 !

Bug fixes : Harder,
More comments, : better,
Major code rework : faster 

A lot of objects (Button handlers, jog wheels, leds) are made reusable for other mappings.
Do you think I should integrate them into common-controllers-script.js

The code was reworked to be closer to the Mixxx file mapping structure philosophy (prototypes, standardized mapping function entry).
JSHint validated (just an orphaned utility function I would like to keep there, for later use may be).

### CODE REVIEW

File: res/controllers/Reloop Beatpad.midi.xml [hunk 1]
```diff
@@ -1153,8 +1133,18 @@
                 </options>
             </control>
             <control>
+                <group>[Channel2]</group>
+                <key>ReloopBeatpad.spflBtn</key>
+                <description>SHIFT+PFL 2</description>
+                <status>0x91</status>
+                <midino>0x21</midino>
```
Comment (sblaisot) [inline res/controllers/Reloop Beatpad.midi.xml]: is it normal that this button's midino switch from 0x61 to 0x21 ?

Comment (sblaisot) [inline res/controllers/Reloop Beatpad.midi.xml]: ok, forget that comment, just saw your changelog in the script ;)

Comment (DJChloe) [inline res/controllers/Reloop Beatpad.midi.xml]: I was about to answer "Yes it's normal" :D

Comment (DJChloe) [inline res/controllers/Reloop Beatpad.midi.xml]: it's consisten with SHIFT+PFL 1 (90/21) , PFL1(90/41)  and PFL2 (91/41)

channel 1 --> 0x90 / channel 2 --> 0x91
shifted control = control - 0x40

**Comment by DJChloe** [comment]
Yop,  I left some some typos aside, a return statement that comes from I don't know where,  and some useless variables defined in the deck object. 
I correct this and go for a new pull request.

**Comment by Be-ing** [comment]
There's no need for a new pull request. Just make another commit to your branch with your edit and push it to your GitHub fork. It will automatically show up in this pull request.

I'll take a look at these changes this week (hopefully I'll have time tonight).

**Comment by DJChloe** [comment]
OK, it's ready.
And this WE we will have a new supported controller in Mixxx, Numark Mixtrack (Pro) 3
The mapping is based on the same reusable code and structure (LEDs / "Super Buttons"/ autoscratch / JogWheels handling). The mapping for the NMTP3 introduces a Fader start feature (purely driven by the mapping).

**Comment by Be-ing** [comment]
Sorry I haven't had time to review this yet. I've been quite busy with work and reviewing the Reloop Jockey 3 ME mapping has taken a lot of time.

I didn't realize that you pretty much rewrote the whole script! Does the mapping still function in all the same ways as described on the wiki page?

The reusable objects sure look interesting. I think it'll probably be useful to move them (or at least some of them) to common-controller-scripts.js, but I want to take a closer look at them and how this script uses them first. @Pegasus-RPG, your input would be appreciated, especially for the Jogger object.

**Comment by DJChloe** [comment]
The scratching effect seems to be awfull with a Numark Mixtrack Pro 3, it
is OK for the reloop beatpad.

I am investigating.

2016-01-19 4:24 GMT+01:00 Be <notifications@github.com>:

**Comment by Be-ing** [comment]
I like the idea of providing an object for managing alternate functionality for buttons depending on whether they are double pressed or long pressed. However, I feel that providing that information as an argument to a single callback function is not the best way to organize the code. I think it would be easier to write code that's easy to manage if each kind of button press (simple, double press, or long press) had its own callback function. I think this could be integrated into the Control framework in PR #1054 in a similar way to how shift functionality is handled. What do you think @DJChloe ?

**Comment by rryan** [comment]
It would be great to have this in Mixxx 2.1 -- @DJChloe  @Be-ing what's left to do here?

**Comment by Be-ing** [comment]
Nothing is left to do as this PR doesn't change any functionality. It's just a refactoring of the code. No problem to merge it as is.

**Comment by rryan** [comment]
Great, thanks @DJChloe

**Pull request merged**